File: python/tracing/exa/03_agent_research_tools.py
from __future__ import annotations

from _shared import clients, example_attributes, finish, make_respan, use_live_exa
from respan import workflow

EXAMPLE = "agent-research-tools"


@workflow(name="exa_python_agent_research_tools")
def agent_research_tools(client):
    search_tool = client.tools.web_search(num_results=1)
    tool_output = search_tool.run({"query": "latest retrieval observability patterns"})
    contents_tool = client.tools.get_contents(text={"max_characters": 500})
    contents_output = contents_tool.run({"urls": ["https://example.com/article"]})

    agent = client.agent.runs.create_and_wait(
        query="Create a one-line brief about retrieval observability.",
        poll_interval=50,
        timeout_ms=120000,
    )

    research_status = "skipped-live-deprecated"
    research_output = None
    if not use_live_exa():
        research = client.research.create(
            instructions="Create a deterministic research brief.",
            model="exa-research-fast",
        )
        completed = client.research.poll_until_finished(
            research.research_id,
            poll_interval=50,
            timeout_ms=10000,
        )
        research_status = completed.status
        research_output = completed.output.content if completed.output else None

    return {
        "tool_output": tool_output,
        "contents_output": contents_output,
        "agent_status": agent.status,
        "agent_output": agent.output.text if agent.output else None,
        "research_status": research_status,
        "research_output": research_output,
    }


def main() -> None:
    respan = make_respan(EXAMPLE)
    try:
        with (
            clients() as (client, _async_client, mode),
            example_attributes(EXAMPLE, mode) as workflow_name,
        ):
            result = agent_research_tools(client)
            print({"workflow_name": workflow_name, "mode": mode, **result})
    finally:
        finish(respan)


if __name__ == "__main__":
    main()

File: typescript/tracing/exa/03_agent_research_tools.ts
import { printResult, runExaExample } from "./_shared.js";

const example = "agent-research-tools";
const result = await runExaExample({
  example,
  fn: async ({ client, mode }) => {
    const tool = client.tools.webSearch({ numResults: 1 });
    const toolOutput = await tool.run({ query: "latest retrieval observability patterns" });
    const agent = await client.agent.runs.createAndWait(
      { query: "Create a one-line brief about retrieval observability." },
      { pollInterval: 50, timeoutMs: 120_000 },
    );

    let researchStatus = "skipped-live-deprecated";
    let researchOutput: unknown;
    if (mode === "loopback") {
      const research = await client.research.create({
        instructions: "Create a deterministic research brief.",
        model: "exa-research-fast",
      });
      const completed = await client.research.pollUntilFinished(research.researchId, {
        pollInterval: 50,
        timeoutMs: 10_000,
      });
      researchStatus = completed.status;
      researchOutput =
        completed.status === "completed"
          ? completed.output
          : { status: completed.status, response: completed };
    }
    return {
      mode,
      toolOutput,
      agentStatus: agent.status,
      agentOutput: agent.output,
      researchStatus,
      researchOutput,
    };
  },
});
printResult(example, result.mode, result);
